proof/invariant-abstract/DetSchedSchedule_AI.thy

```diff
@@ -9459,33 +9495,41 @@ lemma refill_budget_check_round_robin_sc_refills_update_unspecified:
   done
 
 lemma commit_time_released_ipc_queues[wp]:
-  "\<lbrace>released_ipc_queues\<rbrace>
+  "\<lbrace>released_ipc_queues
+    and cur_sc_more_than_ready
+    and active_sc_valid_refills
+    and current_time_bounded 0\<rbrace>
    commit_time
    \<lbrace>\<lambda>_. released_ipc_queues\<rbrace>"
   supply if_split[split del]
   apply (simp add: commit_time_def)
   apply (rule hoare_seq_ext[OF _ gets_sp])
-  apply (rule hoare_seq_ext[OF _ gets_sp])
   apply (rule hoare_seq_ext[OF _ get_sched_context_sp])
   apply (rule hoare_seq_ext[where B="\<lambda>_. released_ipc_queues"], wpsimp)
   apply (rule hoare_when_cases, simp)
+  apply (rule hoare_seq_ext[OF _ gets_sp])
+  apply (rename_tac csc sc consumed)
   apply (rule hoare_seq_ext[where B="\<lambda>_. released_ipc_queues"], wpsimp)
   apply (rule hoare_when_cases, simp)
   apply (rule hoare_seq_ext_skip, solves \<open>wpsimp\<close>, simp?)+
-  apply (rule_tac B="\<lambda>_ s. is_active_sc csc s
-                           \<and> (\<forall>t. pred_map_eq (Some csc) (tcb_scps_of s) t
-                                  \<or> released_ipc_queued_thread t s)"
-           in hoare_seq_ext)
-   apply wpsimp
-   apply (simp add: released_ipc_queues_defs)
-   apply (erule allEI, fastforce simp: obj_at_def vs_all_heap_simps)
-  apply (wpsimp wp: refill_budget_check_sc_refills_update_unspecified
-                    refill_budget_check_round_robin_sc_refills_update_unspecified)
-  apply (rule conjI)
-   apply (clarsimp simp: obj_at_kh_kheap_simps vs_all_heap_simps)
-  apply (simp add: released_ipc_queues_defs)
-  apply (erule allEI)
-  by (fastforce simp: heap_upd_def vs_all_heap_simps)
+  apply (rule hoare_weaken_pre)
+   apply (rule_tac R="cur_sc_more_than_ready and current_time_bounded 0
+                      and (\<lambda>s. consumed = consumed_time s) and cur_sc_active
+                      and (\<lambda>s. valid_refills (cur_sc s) s)"
+          in released_ipc_queues_lift_pre_conj)
+          apply (wpsimp+)[6]
```

Comment:
I'm again a little scared of multiples wpsimps in a row. Are you able to instead have the following wpsimp with a +?

Reply:
That did work. I don't love it, but I'll do it that way since it seems to be the style we're going for.